End of support plugin

mfleeson opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Rewriting the 3.10 End of Service Notification and placing it into Joomla 4 coding structure and style. 
Simplified the language definitions to not be hardcoded for 3 to 4
Currently using const EOS_DATE = '2023-10-25'; in file plugins/quickicon/eos/src/Extension/Eos.php The correct date should be 2025-10-25 but then the notification does not appear. 
Dimitris converted the jQuery code into working javascript for closing the notification and also calling the com_ajax so the hide me is stored in the plugin parameters.


### Testing Instructions
Build. 
Check plg_quickicon_eos is enabled. 
View Home Dashboard in the back end.


### Actual result BEFORE applying this Pull Request
No notification is displayed


### Expected result AFTER applying this Pull Request
A notification is displayed


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

mfleeson committed:
First Draft EOS port from 3.1 - Problem on running

An error has occurred.

    0 Class "Joomla\Plugin\Quickicon\Eos\Extension\Eos" not found
Call stack
# 	Function 	Location
1 	() 	JROOT/plugins/quickicon/eos/services/provider.php:36
2 	Joomla\DI\ServiceProviderInterface@anonymous/home/burninglight/public_html/jvp/plugins/quickicon/eos/services/provider.php:22$f41b->{closure}() 	JROOT/libraries/vendor/joomla/di/src/ContainerResource.php:182
3 	Joomla\DI\ContainerResource->getInstance() 	JROOT/libraries/vendor/joomla/di/src/Container.php:96

mfleeson committed:
Adding Namespace to Eos.xml and fixes to get working correctly.

Adding Namespace to Eos.xml
Fixing issue with Application.
PhpStorm reformat.

mfleeson committed:
Fixing jQuery and Dismiss Notification

Updating the jQuery to close the notification
Debugged the com_ajax code and updated to get the saveParams working.

mfleeson committed:
Updated snooze.js by Dimitri,

Not connecting event to button. Needs updating

dgrammatiko committed:
Update snooze.js

dgrammatiko committed:
Update snooze.js

mfleeson committed:
Merge pull request #1 from dgrammatiko/patch-7

dgrammatiko committed:
Update Eos.php

dgrammatiko committed:
Create snooze.es6.js

dgrammatiko committed:
Delete plugins/quickicon/eos/media directory

dgrammatiko committed:
Update Eos.php

mfleeson committed:
Update eos.xml

Removing media folder as Dimitris has moved the Javascript to the build folder.

dgrammatiko commented on plugins/quickicon/eos/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,21 @@
+; Joomla! Project
```
This file needs to be moved at `administrator/language/en-GB`

dgrammatiko commented on plugins/quickicon/eos/language/en-GB/plg_quickicon_eos.sys.ini:
```diff
@@ -0,0 +1,7 @@
+; Joomla! Project
```
This file needs to be moved at `administrator/language/en-GB`

dgrammatiko commented on plugins/quickicon/eos/services/provider.php:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return  void
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+/**
+ * Joomla! end of support notification plugin
+ *
+ * @since  4.0.0
```
```suggestion
 * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     * The EOS date for 4.x.
+     *
+     * @var    string
+     * @since  4.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     * Load the language file on instantiation.
+     *
+     * @var    bool
+     * @since  4.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     * Holding the current valid message to be shown
+     *
+     * @var    array|bool
+     * @since  4.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @var Document
+     *
+     * @since  4.0.0
```
```suggestion
     * @since  __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  void
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  array|bool  An array with the message to be displayed or false
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  array
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  bool
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  string
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     * @return  array  A list of icon definition associative arrays, consisting of the
+     *                 keys link, image, text and access.
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  bool
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *
+     * @return  bool
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

dgrammatiko commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,361 @@
+     *                                          Recognized key values include 'name', 'group', 'params', 'language'
+     *                                          (this list is not meant to be comprehensive).
+     *
+     * @since   4.0.0
```
```suggestion
     * @since   __DEPLOY_VERSION__
```

sandewt commented on plugins/quickicon/eos/eos.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
```
uppercase
```suggestion
<?xml version="1.0" encoding="UTF-8"?>
```

sandewt commented on plugins/quickicon/eos/eos.xml:
```diff
@@ -0,0 +1,28 @@
+	<files>
+		<folder plugin="eos">services</folder>
+		<folder>src</folder>
+    		<folder>language</folder>
```
Delete
```    		
<folder>language</folder>
```

mfleeson committed:
Changes as per requests

@Since changed to __DEPLOY_VERSION__
utf changed to UTF
language files moved to administrator/language/en-GB
media and language folders deleted from plugin folder and eos.xml updated to remove folders.

mfleeson committed:
Merge branch 'End-of-Support-Plugin' of https://github.com/mfleeson/joomla-cms into End-of-Support-Plugin

mfleeson commented:
All requested changes have been made. Thanks.

dgrammatiko commented on plugins/quickicon/eos/eos.php:
```diff
@@ -1,12 +1,12 @@
 <?php
```
This file is not needed anymore

mfleeson replied:
Deleted it.

sandewt commented on plugins/quickicon/eos/eos.xml:
```diff
@@ -0,0 +1,27 @@
+	<files>
+		<folder plugin="eos">services</folder>
+		<folder>src</folder>
+	</files>
```
These are still needed:
```
<languages>
		<language tag="en-GB">language/en-GB/plg_quickicon_eos.ini</language>
		<language tag="en-GB">language/en-GB/plg_quickicon_eos.sys.ini</language>
</languages>
```

mfleeson replied:
Dimitris said to move those files to administrator/language/en-GB. So do I need them in the plugins/quickicon/eos folder?

mfleeson replied:
Added back into eos.xml

sandewt replied:
> So do I need them in the plugins/quickicon/eos folder?

Yes, in the .xml file.

You will find examples in this folder: https://github.com/mfleeson/joomla-cms/tree/End-of-Support-Plugin/plugins/quickicon

mfleeson committed:
Code Style to keep PHPCS happy

Removed unneeded variable definitions and adjustments needed for PHPCS

mfleeson committed:
Older version of code used as base for src/Extension/Eos.php

Not needed

mfleeson committed:
Fixing Header blocks must be separated by a single blank line

mfleeson committed:
PHPCS fixes

dgrammatiko committed:
Update snooze.es6.js

dgrammatiko committed:
JS CS

heelc29 commented on plugins/quickicon/eos/eos.xml:
```diff
@@ -0,0 +1,27 @@
+	<files>
+		<folder plugin="eos">services</folder>
+		<folder>src</folder>
+	</files>
+	<config>
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+						name="last_snoozed_id"
+						type="hidden"
+				/>
```
```suggestion
				<field
					name="last_snoozed_id"
					type="hidden"
				/>
```

mfleeson replied:
Thanks. In phpstorm it only shows as one tab so i checked in github and edited it back to 1 tab,

mfleeson committed:
Adding back in language definition strings to eos.xml

mfleeson committed:
Merge branch 'End-of-Support-Plugin' of https://github.com/mfleeson/joomla-cms into End-of-Support-Plugin

heelc29 commented on plugins/quickicon/eos/services/provider.php:
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/**
+ * @package                 Joomla.Plugin
+ * @subpackage              Quickicon.Eos
+ * @copyright               (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license                 GNU General Public License version 2 or later; see LICENSE.txt
+ * @phpcs                   :disable PSR1.Classes.ClassDeclaration.MissingNamespace
+ */
```
```suggestion
/**
 * @package     Joomla.Plugin
 * @subpackage  Quickicon.eos
 *
 * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
 * @license     GNU General Public License version 2 or later; see LICENSE.txt
 */
```
format
phpcs should not be necessary

mfleeson replied:
Removed @phpcs, and added back in blank line that was taken out when trying to debug a phpcs error!

heelc29 commented on plugins/quickicon/eos/services/provider.php:
```diff
@@ -0,0 +1,43 @@
+     */
+    public function register(Container $container)
+    {
+        $container->set(PluginInterface::class, function (Container $container) {
+            $dispatcher = $container->get(DispatcherInterface::class);
+            $plugin     = new Eos($dispatcher, Factory::getApplication()->getDocument(), (array)PluginHelper::getPlugin('quickicon', 'eos'));
+            $plugin->setApplication(Factory::getApplication());
+            $plugin->setDatabase($container->get(DatabaseInterface::class));
+
+            return $plugin;
+        });
```
```suggestion
        $container->set(
            PluginInterface::class,
            function (Container $container) {
                $dispatcher = $container->get(DispatcherInterface::class);
                $plugin     = new Eos(
                    $dispatcher,
                    Factory::getApplication()->getDocument(),
                    (array) PluginHelper::getPlugin('quickicon', 'eos')
                );
                $plugin->setApplication(Factory::getApplication());
                $plugin->setDatabase($container->get(DatabaseInterface::class));

                return $plugin;
            }
        );
```
format like other plugins

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,356 @@
+<?php
+
+/**
+ * @package                 Joomla.Plugin
+ * @subpackage              quickicon.eos
+ * @copyright               (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license                 GNU General Public License version 2 or later; see LICENSE.txt
+ * @phpcs                   :disable PSR1.Classes.ClassDeclaration.MissingNamespace
+ */
```
```suggestion
/**
 * @package     Joomla.Plugin
 * @subpackage  Quickicon.eos
 *
 * @copyright   (C) 2023 Open Source Matters, Inc. <https://www.joomla.org>
 * @license     GNU General Public License version 2 or later; see LICENSE.txt
 */
```
format
file has namespace

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,21 @@
+PLG_QUICKICON_EOS_MESSAGE_INFO_01_SHORT="The latest version of Joomla has arrived! <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_02="<p>When is the time to migrate to the next version? Once the extensions your site needs are compatible. Learn <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">how to use the Pre-Update Checker</a>. <span class=\"icon-new-tab\"></span></p>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_02_SHORT="Use Pre-Update Check for extension compatibility <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_WARNING_SECURITY_ONLY="<p>This version of Joomla has entered security only mode. Support ends %1$s. Start <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">planning to migrate</a> <span class=\"icon-new-tab\"></span>to Joomla 4 today.</p>"
```
Joomla 4?

mfleeson replied:
I did say in Mattermost that the message text needed changing. Especially the documentation links.

heelc29 replied:
Ok, because all other strings about Joomla 3.10 and 4 were replaced ...

mfleeson committed:
Formatting corrections

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,21 @@
+; License GNU General Public License version 2 or later; see LICENSE.txt
+; Note : All ini files need to be saved as UTF-8
+
+PLG_QUICKICON_EOS="Quick Icon - Joomla End Of Support Notification"
```
```suggestion
PLG_QUICKICON_EOS="Quick Icon - Joomla! End Of Support Notification"
```
to be in sync with sys file

mfleeson committed:
Update eos.xml

mfleeson committed:
Update Eos.php

mfleeson committed:
Merge branch 'End-of-Support-Plugin' of https://github.com/mfleeson/joomla-cms into End-of-Support-Plugin

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+     * @var    array|bool
+     * @since __DEPLOY_VERSION__
+     */
+    private array|bool $currentMessage = [];
```
```suggestion
    private $currentMessage = [];
```
not allowed in PHP7

mfleeson replied:
All my JED 4 coding has been in php 8 and 8.1. PHPStorm recommended these. I've removed them.

sandewt commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,356 @@
+/**
+ * Joomla! end of support notification plugin
+ *
+ * @since __DEPLOY_VERSION__
```
Do not use `__DEPLOY_VERSION__` everywhere here, but leave the original date if necessary.
Take a look at this file as an example: https://github.com/mfleeson/joomla-cms/blob/End-of-Support-Plugin/plugins/quickicon/phpversioncheck/src/Extension/PhpVersionCheck.php

[EDIT Or have these files not been used before?]

mfleeson replied:
ok. so just at the header and then fixed to 4.4.0 as thats when it will release?

sandewt replied:
> ok. so just at the header and then fixed to 4.4.0 as thats when it will release?

No, I didn't mean it that way.

For example, if the constante / function / method has been used before in J3.10, the version remains unchanged.

See line 104 - 106:
```
* @since   3.10.0
*/
private function getMessageInfo($monthsUntilEOS, $inverted)
{
```
So not: `@since   4.4.0`

sandewt replied:
@mfleeson 

Sorry, I wasn't clear with my comment.
Anything **new** will be `__DEPLOY_VERSION__` and NOT `4.4.0`

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+     *
+     * @since __DEPLOY_VERSION__
+     */
+    private function getMessageInfo(int $monthsUntilEOS, int $inverted): bool|array
```
```suggestion
    private function getMessageInfo(int $monthsUntilEOS, int $inverted)
```
not allowed in PHP7

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+                    $cachecontroller = new CacheController($options);
+                    $cache           = $cachecontroller->cache;
+                    $cache->clean();
+                } catch (Exception) {
```
```suggestion
                } catch (Exception $e) {
```
not allowed in PHP7

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+        try {
+            // Lock the tables to prevent multiple plugin executions causing a race condition
+            $db->lockTable('#__extensions');
+        } catch (Exception) {
```
```suggestion
        } catch (Exception $e) {
```
not allowed in PHP7

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+            // Update the plugin parameters
+            $result = $db->setQuery($query)->execute();
+            $this->clearCacheGroups();
+        } catch (Exception) {
```
```suggestion
        } catch (Exception $e) {
```
not allowed in PHP7

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,357 @@
+        try {
+            // Unlock the tables after writing
+            $db->unlockTables();
+        } catch (Exception) {
```
```suggestion
        } catch (Exception $e) {
```
not allowed in PHP7

richard67 commented:
Complete SQL stuff for new installations and updates is missing, and the plugin is also missing in the extensions helper.

mfleeson committed:
Removing php8 code and making it 7 compatible and adding php3.10 strings for reference to language

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,21 @@
+PLG_QUICKICON_EOS_GROUPNAME_WARNING="Support Status"
+PLG_QUICKICON_EOS_MESSAGE_ERROR_SUPPORT_ENDED="<p>Support has ended for your version of Joomla. <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">Migrate to the next version</a> <span class=\"icon-new-tab\"></span>as soon as possible.</p>"
+PLG_QUICKICON_EOS_MESSAGE_ERROR_SUPPORT_ENDED_SHORT="Support has ended for your version of Joomla <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_01="<p>The latest version of Joomla has arrived! Find out all that it has to offer you. Check the landing page for <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">its features</a> <span class=\"icon-new-tab\"></span>and improvements.</p>"
```
```suggestion
PLG_QUICKICON_EOS_MESSAGE_INFO_01="<p>The latest version of Joomla has arrived! Find out all that it has to offer you. Check the landing page for <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">its features</a> and improvements.</p>"
```
![image](https://user-images.githubusercontent.com/66922325/236694269-d981748d-7abd-4a2b-87d8-fbfdaf7ac52f.png)

mfleeson replied:
I did wonder about that. It was lifted straight from 3.10

heelc29 replied:
Maybe its new in Joomla 4 🤔
https://github.com/joomla/joomla-cms/blob/be4f48dd66da1dce0d2a1a9acb325cce1d1ebb06/build/media_source/templates/administrator/atum/scss/blocks/_global.scss#L178-L185

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,21 @@
+PLG_QUICKICON_EOS_MESSAGE_ERROR_SUPPORT_ENDED_SHORT="Support has ended for your version of Joomla <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_01="<p>The latest version of Joomla has arrived! Find out all that it has to offer you. Check the landing page for <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">its features</a> <span class=\"icon-new-tab\"></span>and improvements.</p>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_01_SHORT="The latest version of Joomla has arrived! <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_02="<p>When is the time to migrate to the next version? Once the extensions your site needs are compatible. Learn <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">how to use the Pre-Update Checker</a>. <span class=\"icon-new-tab\"></span></p>"
```
```suggestion
PLG_QUICKICON_EOS_MESSAGE_INFO_02="<p>When is the time to migrate to the next version? Once the extensions your site needs are compatible. Learn <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">how to use the Pre-Update Checker</a>.</p>"
```

![image](https://user-images.githubusercontent.com/66922325/236694363-02b66838-06ef-4e9d-a4de-75ee0be8af15.png)

mfleeson committed:
Update plg_quickicon_eos.ini

mfleeson commented:
Richard has informed me I need to include it in the extensions table and the ExtensionHelper so I'll take a look at that tomorrow and then rerequest the PR.

heelc29 commented on plugins/quickicon/eos/services/provider.php:
```diff
@@ -0,0 +1,50 @@
+                    (array)PluginHelper::getPlugin('quickicon', 'eos')
+                );
+                $plugin->setApplication(Factory::getApplication());
+                $plugin->setDatabase($container->get(DatabaseInterface::class));
```
```suggestion
```

heelc29 commented on plugins/quickicon/eos/services/provider.php:
```diff
@@ -0,0 +1,50 @@
+use Joomla\CMS\Extension\PluginInterface;
+use Joomla\CMS\Factory;
+use Joomla\CMS\Plugin\PluginHelper;
+use Joomla\Database\DatabaseInterface;
```
```suggestion
```

mfleeson replied:
That's already been removed.

mfleeson replied:
https://github.com/mfleeson/joomla-cms/commit/0572c8b8f115a948e5084f7598c199f90567458a

heelc29 replied:
Only from extension file but not from provider, so no it's broken
![image](https://user-images.githubusercontent.com/66922325/236695213-0f3c1f3e-8730-4ebf-b502-d7a8e43e611a.png)

heelc29 replied:
But database is still used here:
https://github.com/joomla/joomla-cms/blob/369c3a383d43e4ff44ccfd2c4b2f772c25055302/plugins/quickicon/eos/src/Extension/Eos.php#L288

heelc29 replied:
Only from extension file but not from provider, so now it's broken
![image](https://user-images.githubusercontent.com/66922325/236695468-27973650-b862-4e48-a5d3-cdf734abd410.png)

but the database is used here
https://github.com/joomla/joomla-cms/blob/369c3a383d43e4ff44ccfd2c4b2f772c25055302/plugins/quickicon/eos/src/Extension/Eos.php#L288

heelc29 reviewed (commented):
since commit 0572c8b8f115a948e5084f7598c199f90567458a

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,38 @@
+PLG_QUICKICON_EOS_MESSAGE_INFO_01_SHORT="The latest version of Joomla has arrived! <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_02="<p>When is the time to migrate to the next version? Once the extensions your site needs are compatible. Learn <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">how to use the Pre-Update Checker</a>. </p>"
+PLG_QUICKICON_EOS_MESSAGE_INFO_02_SHORT="Use Pre-Update Check for extension compatibility <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_WARNING_SECURITY_ONLY="<p>This version of Joomla has entered security only mode. Support ends %1$s. Start <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">planning to migrate</a> <span class=\"icon-new-tab\"></span>to the next version of Joomla today.</p>"
```
```suggestion
PLG_QUICKICON_EOS_MESSAGE_WARNING_SECURITY_ONLY="<p>This version of Joomla has entered security only mode. Support ends %1$s. Start <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">planning to migrate</a> to the next version of Joomla today.</p>"
```
![image](https://user-images.githubusercontent.com/66922325/236695004-f3f09742-bdf9-4890-a975-6e0103e48071.png)

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,38 @@
+PLG_QUICKICON_EOS_MESSAGE_INFO_02_SHORT="Use Pre-Update Check for extension compatibility <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_WARNING_SECURITY_ONLY="<p>This version of Joomla has entered security only mode. Support ends %1$s. Start <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">planning to migrate</a> <span class=\"icon-new-tab\"></span>to the next version of Joomla today.</p>"
+PLG_QUICKICON_EOS_MESSAGE_WARNING_SECURITY_ONLY_SHORT="This version of Joomla ends support on %1$s. <span class=\"icon-new-tab\"></span>"
+PLG_QUICKICON_EOS_MESSAGE_WARNING_SUPPORT_ENDING="<p>Support ends on %1$s for this version of Joomla. <span class=\"icon-new-tab\"></span><a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">Migrate to the latest version of Joomla</a> as soon as possible.</p>"
```
```suggestion
PLG_QUICKICON_EOS_MESSAGE_WARNING_SUPPORT_ENDING="<p>Support ends on %1$s for this version of Joomla. <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">Migrate to the latest version of Joomla</a> as soon as possible.</p>"
```
![image](https://user-images.githubusercontent.com/66922325/236695062-4672849d-8e90-4886-a269-5e45ea501633.png)

heelc29 commented on administrator/language/en-GB/plg_quickicon_eos.ini:
```diff
@@ -0,0 +1,38 @@
+PLG_QUICKICON_EOS_GROUPNAME_EOS="End Of Support"
+PLG_QUICKICON_EOS_GROUPNAME_INFO="Upgrade Information"
+PLG_QUICKICON_EOS_GROUPNAME_WARNING="Support Status"
+PLG_QUICKICON_EOS_MESSAGE_ERROR_SUPPORT_ENDED="<p>Support has ended for your version of Joomla. <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">Migrate to the next version</a> <span class=\"icon-new-tab\"></span>as soon as possible.</p>"
```
```suggestion
PLG_QUICKICON_EOS_MESSAGE_ERROR_SUPPORT_ENDED="<p>Support has ended for your version of Joomla. <a href=\"%2$s\" target=\"_blank\" rel=\"noopener noreferrer\">Migrate to the next version</a> as soon as possible.</p>"
```
![image](https://user-images.githubusercontent.com/66922325/236695278-bb61c03d-e315-46f8-a759-18ba2298a9a9.png)

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,355 @@
+     *
+     * @throws Exception
+     */
+    public function onGetIcons(string $context): array
```
Maybe check the context?
![image](https://user-images.githubusercontent.com/66922325/236695646-b96210f0-48fd-4431-9c5a-2009a7d71fca.png)

heelc29 commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,355 @@
+     */
+    private $currentMessage = [];
+
+    /**
+     * The document.
+     *
+     * @var Document
+     *
+     * @since 4.4.0
+     */
+    private Document $document;
```
Document is available from application ... so is the extra variable (in constructor) necessary?
https://github.com/joomla/joomla-cms/blob/be4f48dd66da1dce0d2a1a9acb325cce1d1ebb06/libraries/src/Factory.php#L324-L326

mfleeson replied:
Currently the code passes Document from provider.php on line 40. This is based on the existing plugins/quickicon/joomlaupdate/servers/provider.php Should this be different?

sandewt commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,355 @@
+
+namespace Joomla\Plugin\Quickicon\Eos\Extension;
+
+// phpcs:disable PSR1.Files.SideEffects
+defined('_JEXEC') || die;
+
+// phpcs:enable PSR1.Files.SideEffects
```
Use:
```
// phpcs:disable PSR1.Files.SideEffects
\defined('_JEXEC') or die;
// phpcs:enable PSR1.Files.SideEffects
```

mfleeson replied:
Not sure why phpstorm added a blank line after defined, have added a \ to defined.

mfleeson replied:
though the example i was using joomlaupdate\services\provider.php does not have a \ before the defined

mfleeson replied:
and neither does phpversioncheck

sandewt replied:
_I usually look at what is most common within Joomla to be consistent._ The results of PHPStorm sometimes deviate. When in doubt, delve into history. If necessary, ask a question about, in the hope a get an answer. Or finally enter it as a new issue. 

It is a continuous learning process for me. :sweat_smile:

brianteeman committed:
[4.3] codestyle indentation (#40551)

simple pr - code review only

dgrammatiko committed:
[4.3] Codemirror corrects path for keybinding script +(obey debug - versioning) (#40051)

janschoenherr committed:
Prevent Exception (#40122)

* Prevent Exception

If $parts is `null` this will throw an exception.

```count(): Argument #1 ($value) must be of type Countable|array, null given```

This can happen if the context is e.g. the default context `text` from the `Joomla\CMS\HTML\Helpers\Content::prepare()` function.

* Update fields.php

dgrammatiko committed:
[4.3] Media manager, refactor the session fallback (#40013)

mfleeson committed:
Limit notification to one group, adding wider install sql/ExtensionHelper

Adding plugin to #__extensions in base and update sql
Adding plugin to ExtensionHelper.php
Updating Eos.php to limit notification to just Notification panel on home

mfleeson committed:
Merge branch '4.4-dev' into End-of-Support-Plugin

mfleeson committed:
Renaming upgrade sql and adding postgres upgrade sql

mfleeson committed:
Merge branch 'End-of-Support-Plugin' of https://github.com/mfleeson/joomla-cms into End-of-Support-Plugin

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/4.4.0-2023-05-08.sql:
```diff
@@ -0,0 +1,2 @@
+INSERT INTO "#__extensions" ("package_id", "name", "type", "element", "folder", "client_id", "enabled", "access", "protected", "locked", "manifest_cache", "params", "custom_data", "ordering", "state") VALUES
+	(0, "plg_quickicon_eos", "plugin", "eos", "quickicon", 0, 1, 1, 0, 1, "", "", "", 7, 0);
```
```suggestion
	(0, 'plg_quickicon_eos', 'plugin', 'eos', 'quickicon', 0, 1, 1, 0, 1, '', '', '', 7, 0);
```
Only name quotes are double quotes in PostgreSQL, string quotes are single quotes like in MySQL.

richard67 replied:
P.S.: That's not just code style, it's a syntax error.

mfleeson replied:
Thanks. Not worked with Postgres for about 25 years.

mfleeson committed:
Updating @since and fixing a syntax error in postgres update.

mfleeson committed:
phpcs fix

mfleeson committed:
One extra space! now deleted

sandewt commented on plugins/quickicon/eos/eos.xml:
```diff
@@ -0,0 +1,31 @@
+	<license>GNU General Public License version 2 or later; see LICENSE.txt</license>
+	<authorEmail>admin@joomla.org</authorEmail>
+	<authorUrl>www.joomla.org</authorUrl>
+	<version>4.0.0</version>
```
I think this should be `<version>3.10.0</version>` wich corresponds to the original eos.xml file.
But I doubt. Maybe someone else has advice?

sandewt replied:
PS. The eos.php file also has references to 3.10.0

mfleeson replied:
Although the base code is from 3.10 the whole codebase was restructured to use 4.x standards so technically this code is at least 4.x based. Some of the functions were left using the 3.10 logic but still were moved to 4.x.

sandewt replied:
> Although the base code is from 3.10 the whole codebase was restructured to use 4.x standards so technically this code is at least 4.x based. Some of the functions were left using the 3.10 logic but still were moved to 4.x.

I already understood this, hence my comment.
All the functions (functionality) were left using the 3.10.0, then refer to 3.10.0. As far as I understood.

Ping @laoneo

richard67 replied:
The code was removed from the 4.0-dev branch once, so the plugin introduced here is something new, even if the code is based on the old one from 3.10-dev. The old one was named "plg_quickicon_eos310" and is uninstalled when updating to J4.x, see https://github.com/joomla/joomla-cms/blob/4.3-dev/administrator/components/com_admin/script.php#L491 . So all code here should get the `__DEPLOY_VERSION__` (or 4.4.0 in the XML), there should no "since" tag or so be left from 3.10.

sandewt replied:
@richard67 

Thank you for your clear and complete explanation. I totally agree with it. 👍

dgrammatiko replied:
Funny part I had raised suggestions for all of them (ie: https://github.com/joomla/joomla-cms/pull/40556#discussion_r1186869206) but probably a forced pushed reverted all of them

sandewt replied:
> Funny part I had raised suggestions for all of them (ie: https://github.com/joomla/joomla-cms/pull/40556#discussion_r1186869206) but probably a forced pushed reverted all of them

MEA culpa

sandewt commented on plugins/quickicon/eos/src/Extension/Eos.php:
```diff
@@ -0,0 +1,382 @@
+namespace Joomla\Plugin\Quickicon\Eos\Extension;
+
+// phpcs:disable PSR1.Files.SideEffects
+\defined('_JEXEC') || die;
```
Common is: _or_
```suggestion
\defined('_JEXEC') or die;
```

mfleeson committed:
Merge branch 'joomla:4.3-dev' into End-of-Support-Plugin

richard67 commented:
Why closed PR and deleted branch?

mfleeson commented:
I'm sorry I didn't realise it had done that. I thought I was working on my local fork. I was trying to get all of the files back into one place. I didn't delete the branch, just renamed it as I realised it was based on 4.3 and not 4.4. So I created a new fork based on 4.4 and have used a comparison to put all of the code into the new fork. I was going to make sure it adheres to all of these comments and then push that. Git is still very much a black art to me.

richard67 commented:
Then I think you have to make a new PR with the renamed branch. If you do that, please refer to this PR here in the description of the new PR so people can find previous discussions.

mfleeson commented:
I was using github desktop, that's the scary thing but it wasnt updating with other peoples changes so it totally overwrote Dimitri's javascript code. I think I've got it all back in hand now and am going to try and create a new PR for 4.4

The new PR is https://github.com/joomla/joomla-cms/pull/40565

sandewt commented:
> I was using github desktop, that's the scary thing but it wasnt updating with other peoples changes so it totally overwrote Dimitri's javascript code.

Thanks for the info. Good to know.
